Loading BitmapFonts at runtime

agaertner opened the pull request:
Using SpriteFontPlus.Monogame we can load and bake TrueTypeFonts at runtime and convert them to a Monogame.Extended.BitmapFont using the exposed texture.

Discussion ref: https://discord.com/channels/531175899588984842/536970543736291346/1147888419364687892

## Is this a breaking change?
No

agaertner committed:
feat: drop in FontStashSharp behind Extended.BitmapFont

agaertner committed:
fix: dispose fontsystem after bitmapfont conversion

agaertner committed:
fix: use spriteFontPlus as it exposes all glyphs

agaertner committed:
refactor: to extension

agaertner committed:
refactor: maintain code style

dlamkins commented on Blish HUD/GameServices/Modules/Managers/ContentsManager.cs:
```diff
@@ -93,11 +94,25 @@ public class ContentsManager : IDisposable {
+
+            if (fontDataLength > 0) {
+                using var ctx = GameService.Graphics.LendGraphicsDeviceContext();
+                var bakeResult = TtfFontBaker.Bake(fontData, (int)size, 1024, 1024, new[] {
```
Do we have any indication of how much can fit on each of these render surfaces?  Is 1024 chosen arbitrarily, and what range of sizes can it hold if we're putting all Latin, Latin1Supplement, Latin extended, and Cyrillic on it?  😅

agaertner replied:
It's taken from the example in <https://github.com/rds1983/SpriteFontPlus> for size 25. 
It should hold anything up to 25 otherwise we can increase its size. I am myself am not sure how many fit on one of these but I can make some tests

agaertner replied:
**Tests**

- Maximum FontSize.36
- Removed CharacterRange.Cyrillic
- Fonts: Lato-Regular.ttf, PTSerif-Regular, GWTwoFont, Cagliostro-Regular.ttf, StoweOpenFace

PTSerif-Regular: Working good.
![image](https://github.com/blish-hud/Blish-HUD/assets/13819164/12a5ef89-5dfb-4a51-a359-5cecc32d9ea9)

 GWTwoFont: Working good.
![image](https://github.com/blish-hud/Blish-HUD/assets/13819164/8c026cef-cb69-41bb-9d53-8226b57bd9bc)

Cagliostro-Regular: Working good
![image](https://github.com/blish-hud/Blish-HUD/assets/13819164/40757e4a-dda8-4000-ae9d-cf1ab61f5144)

StoweOpenFace: Line height seems wrong. But that might be an issue with the font itself or since it is a title font the line height is supposed to overlay a bit by design.
![image](https://github.com/blish-hud/Blish-HUD/assets/13819164/b13fb121-c146-4dc5-8bbf-67f4d85b9bf8)

Result: Added optional line height parameters for "wonky" font line heights. I expect none to be wonky like StowOpenFace. It's likely by design. But able to set a custom line height is good anyway.

Commit: https://github.com/blish-hud/Blish-HUD/pull/913/commits/1ec1012370d5a80463ab2e56fd2431b4bf33e6e8

agaertner committed:
fix: remove cyrillic characters and add optional line height param

agaertner committed:
skip: add overload for fontSize param as int

agaertner committed:
fix: line height exception conditional

agaertner committed:
fix: hook up custom lineHeight param

agaertner committed:
fix: remove FontSize overload because misleading

agaertner committed:
fix: add optional textureSize param

agaertner committed:
fix: pass textureSize in GetBitmapFont

agaertner committed:
fix: make public

Avoid duplicate font baking and SpriteFont loading when mixing native with non-native draw calls.

agaertner committed:
fix: extend BitmapFont with IDisposable to unload module fonts

agaertner committed:
fix: ambiguity between Extended.BitmapFont and Content.BitmapFont

agaertner commented on Blish HUD/Controls/_Types/FormattedLabelPart.cs:
```diff
@@ -1,8 +1,7 @@
-using Microsoft.Xna.Framework.Graphics;
-using MonoGame.Extended.BitmapFonts;
+using System;
+using BitmapFont = MonoGame.Extended.BitmapFonts.BitmapFont;
```
Fix ambiguity between Extended.BitmapFont and Content.BitmapFont.

agaertner commented on Blish HUD/Controls/_Types/CachedStringRender.cs:
```diff
@@ -1,12 +1,9 @@
-using MonoGame.Extended.BitmapFonts;
+using System;
+using System.Collections.Concurrent;
+using BitmapFont = MonoGame.Extended.BitmapFonts.BitmapFont;
```
Fix ambiguity between Extended.BitmapFont and Content.BitmapFont.

agaertner commented on Blish HUD/GameServices/Content/BitmapFont.cs:
```diff
@@ -0,0 +1,42 @@
+    /// <summary>
+    /// Extends <see cref="MonoGame.Extended.BitmapFonts.BitmapFont"/> to allow disposing of its glyph lookup texture.
+    /// </summary>
+    public class BitmapFont : MonoGame.Extended.BitmapFonts.BitmapFont, IDisposable {
```
Textures of fonts loaded via modules should be disposed when a module unloads.
This will extend Extended.BitmapFont to allow for disposal of the glyph lookup texture.
Safe: After disposal the font just won't draw anymore and will **not** throw.
Also important for "font size" user settings to dispose old sizes and free memory.

agaertner commented on Blish HUD/GameServices/ContentService.cs:
```diff
@@ -1,14 +1,14 @@
+using System.IO;
+using System.IO.Compression;
+using System.Text.RegularExpressions;
+using BitmapFont = MonoGame.Extended.BitmapFonts.BitmapFont;
```
Fix ambiguity between Extended.BitmapFont and Content.BitmapFont.

agaertner committed:
fix: support gw2 known character ranges

agaertner committed:
Merge branch 'dev' into feature/runtime-fonts

agaertner committed:
Merge branch 'dev' into feature/runtime-fonts

agaertner committed:
refactor: rename BitmapFont -> BitmapFontEx; revert usings
